firedrake/slate/slate.py

```diff
@@ -495,6 +511,8 @@ class Block(TensorBase):
     """
 
     def __new__(cls, tensor, indices):
+        assert tensor.terminal, "Slate can only express Blocks on terminal tensors."
```

Comment:
Is this a restriction we want to impose?

Reply:
I **think** we don't have any tests that test blocks on nodes which are not terminal. But I have just double-checked and blocks on non-terminal tensors do work on master, so we probably don't want to impose this restriction here.

Reply:
Idea:

Don't require terminal tensors here, but, have an index-inliner pass in `compile_slate` that does something like:
```
Block(tensor, indices) -> subtensor if tensor.terminal else Block(tensor, indices)
```

Reply:
There is a draft of the optimisation pass in https://github.com/firedrakeproject/firedrake/pull/2111/commits/562f7d64d53d35bf21dce825b191758ee642e54a. It's passing some tests for blocks on 2-tensors. I will need to add more testing for blocks on 1-Tensors and AssembledVectors, expressions involving inverse/solve/factorisation, one expression which is a bit more involved, and expressions which do not only involve one tensor I guess.

How does the design of the overall optimisation pass look, though? Any comments?